FormalSystems/Computability/Loop.lean

```diff
@@ -22,18 +22,52 @@ inductive LoopProgram
 | loop (x: Variable) (p: LoopProgram)
 deriving DecidableEq
 
-def LoopProgram.toFunction (p: LoopProgram):
-  Nat → Nat :=
-  -- Todo: define how to evaluate `p`, with x₀ bound to the input
-  -- for example one could use a state monad with an `AssocList`
-  sorry
-
 def LoopProgram.len: LoopProgram → Nat
 | add x y n => x.index + y.index + n + 1
 | sub x y n => x.index + y.index + n + 1
 | concat p₁ p₂ => p₁.len + p₂.len + 1
 | loop n p => p.len + n.index + 1
 
+def LoopProgramState := Lean.AssocList Variable Nat 
+
+namespace LoopProgramState
+  def get (s: LoopProgramState) (v: Variable): Nat := (s.find? v).getD 0
+
+  def put (s: LoopProgramState) (v: Variable) (n: Nat): LoopProgramState := 
+    if s.contains v then s.replace v n else s.insert v n
+end LoopProgramState
+
+mutual
+  def LoopProgram.runNTimes (p: LoopProgram) (n: Nat) (s: LoopProgramState): LoopProgramState :=
```

Comment:
Maybe instead of defining this yourself, you could use `Nat.iterate`?

Reply:
Ah nice, I was looking for something like this :)